src/irr/asset/COBJMeshFileLoader.cpp

```diff
@@ -39,13 +39,361 @@ namespace std
+        {
+            if (PC.roughness==0.0)
+                diffuse = 1.0/PI;
+            else
+                diffuse = oren_nayar(PC.roughness, N, V, V, NdotL, NdotV);
+        }
+        float specular = 0.0;
+        if (NdotV > FLT_MIN)
+        {
+            float ndf = GGXBurleyAnisotropic(PC.anisotropy, PC.roughness, TdotH, BdotH, NdotH);
+            float geom = GGXSmithHeightCorrelated_aniso_wo_numerator(at, ab, TdotL, TdotV, BdotL, BdotV, NdotL, NdotV);
+            specular = ndf*geom*fr;
+        }
+
+        color += (diffuseFactor*diffuse*PC.albedo + specular) * NdotL / lightDistance2;
+    }
+    OutColor = vec4(color*PC.transmissionFilter, 1.0);
+}
+)";
+    constexpr const char* VERT_SHADER_UV = 
+R"(#version 430 core
+
+layout (location = 0) in vec3 vPos;
+layout (location = 2) in vec2 vUV;
+layout (location = 3) in vec3 vNormal;
+
+layout (location = 0) out vec3 LocalPos;
+layout (location = 1) out vec3 ViewPos;
+layout (location = 2) out vec3 Normal;
+layout (location = 3) out vec2 UV;
+
+layout (set = 1, binding = 0, row_major, std140) uniform UBO {
+    mat4 MVP;
+    mat4 MV;
+    mat3 NormalMat;
+    vec3 EyePos;
+} CamData;
+
+void main()
+{
+    LocalPos = vPos;
+    gl_Position = CamData.MVP*vec4(vPos, 1.0);
+    ViewPos = (CamData.MV*vec4(vPos, 1.0)).xyz;
+    Normal = normalize(CamData.NormalMat*vNormal);
+    UV = vUV;
+}
+)";
+    constexpr const char* FRAG_SHADER_UV =
+R"(#version 430 core
+
+layout (location = 0) in vec3 LocalPos;
+layout (location = 1) in vec3 ViewPos;
+layout (location = 2) in vec3 Normal;
+layout (location = 3) in vec2 UV;
+layout (location = 0) out vec4 OutColor;
+
+#define ILLUM_MODEL_MASK 0x0fu
+#define map_Ka_MASK uint(1u<<4u)
+#define map_Kd_MASK uint(1u<<5u)
+#define map_Ks_MASK uint(1u<<6u)
+#define map_Ns_MASK uint(1u<<8u)
+#define map_d_MASK uint(1u<<9u)
+#define map_bump_MASK uint(1u<<10u)
+#define map_normal_MASK uint(1u<<11u)
+
+layout (push_constant) uniform Block {
+    vec3 Ka;
+    vec3 Kd;
+    vec3 Ks;
+    vec3 Ke;
+    vec4 Tf;//w component doesnt matter
+    float Ns;
+    float d;
+    float bm;
+    float Ni;
+    float roughness;
+    float metallic;
+    float sheen;
+    float clearcoatThickness;
+    float clearcoatRoughness;
+    float anisotropy;
+    float anisoRotation;
+    //extra info
+    uint extra;
+} PC;
+
+//here texture bindings will be inserted with sprintf()
+%s
+
+#include <irr/builtin/glsl/brdf/specular/fresnel/fresnel.glsl>
+
+#define Ia 0.1
+void main()
+{
+    vec3 N = normalize(Normal);
+    if ((PC.extra&map_bump_MASK) == map_bump_MASK)
+    {
+        float height = texture(map_bump, UV).x;
+        vec3 dpdx = dFdx(ViewPos);
+        vec3 dpdy = dFdy(ViewPos);
+        float dhdx = dFdx(height);
+        float dhdy = dFdy(height);
+        
+        vec3 r1 = cross(dpdy, N);
+        vec3 r2 = cross(N, dpdx);
+        vec3 surfGrad = (r1*dhdx + r2*dhdy) / dot(dpdx,r1);
+        N = normalize(N - surfGrad);
+    }
```

Comment:
derivative mapping and bump mapping functions should be in common include https://github.com/buildaworldnet/IrrlichtBAW/issues/384#issuecomment-546903906

P.S. Also support DOT3 normal maps for OBJs